server/src/user/service/user.service.ts

```diff
@@ -192,4 +192,40 @@ export class UserService {
 
     await this.userRepository.save(user);
   }
+
+  async requestPasswordReset(email: string) {
+    const user = await this.userRepository.findOne({
+      where: { email: email },
+    });
+
+    if (!user) {
+      return;
+    }
+
+    const uuid = uuidv4();
+    await this.redisService.set(
+      `${REDIS_KEYS.USER_RESET_PASSWORD_KEY}:${uuid}`,
+      JSON.stringify(user),
+      'EX',
+      600,
+    );
+
+    this.emailService.sendPasswordResetEmail(user, uuid);
+  }
+
+  async resetPassword(uuid: string, password: string): Promise<void> {
+    const userData = await this.redisService.get(
+      `${REDIS_KEYS.USER_RESET_PASSWORD_KEY}:${uuid}`,
+    );
+
+    if (!userData) {
+      throw new NotFoundException('인증에 실패했습니다.');
+    }
+
+    const user = JSON.parse(userData);
+    user.password = await this.createHashedPassword(password);
+
+    this.redisService.del(`${REDIS_KEYS.USER_RESET_PASSWORD_KEY}:${uuid}`);
+    await this.userRepository.save(user);
```

Comment:
```suggestion
    const userData = await this.redisService.get(
      `${REDIS_KEYS.USER_RESET_PASSWORD_KEY}:${uuid}`,
    );

    if (!userData) {
      throw new NotFoundException('인증에 실패했습니다.');
    }
    const user = // 고유값으로 유저 DB에서 조회

    
    user.password = await this.createHashedPassword(password);
    this.redisService.del(`${REDIS_KEYS.USER_RESET_PASSWORD_KEY}:${uuid}`);
    await this.userRepository.save(user);
```
P2: 현재 API의 목적은 비밀번호 변경이기에, stale한 데이터를 들고와서 직렬화 후 DB에 넣기보다는 새롭게 DB에서 fresh하게 조회한 후 유저 객체를 저장하는 것이 좋을 것 같아요. 

그럴 가능성은 적겠지만, 만약에 사용자가 비밀번호 변경을 요청하고 인증을 수행하는 사이에 유저 정보가 뭔가 변경된다면, 이 API로 인해서 그 변경이 이전 상태로 돌아갈 수 있어요.

Reply:
윗 부분과 연관성이 있는 부분이네요.

먼저 제가 로직을 작성한 이유는 레디스를 인증을 위한 세션에 더해서 캐싱 기능까지 같이 생각을 했어요.
`미리 유저 정보를 직렬화해서 캐싱한다면 굳이 DB를 한 번 더 체크할 필요가 없지 않을까?`라는 질문이 떠올랐고,
코드로 옮긴 게 성윤님이 짚어준 두 개의 코드 블럭이고요.

> 그럴 가능성은 적겠지만, 만약에 사용자가 비밀번호 변경을 요청하고 인증을 수행하는 사이에 유저 정보가 뭔가 변경된다면, 이 API로 인해서 그 변경이 이전 상태로 돌아갈 수 있어요.

이 부분은 미처 생각을 못했네요.
안전하게 DB를 한 번 더 체크해보는 게 맞는 것 같습니다. 
좋은 예시 감사합니다.👍